File: src/main/java/dev/kensa/util/IoUtil.java
package dev.kensa.util;

import dev.kensa.KensaException;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;

import static java.nio.file.FileVisitResult.CONTINUE;

public final class IoUtil {

    private static final DeleteAllVisitor DELETE_ALL_VISITOR = new DeleteAllVisitor();

    public static void recreate(Path path) {
        try {
            if (Files.exists(path)) {
                Files.walkFileTree(path, DELETE_ALL_VISITOR);
            }
            Files.createDirectories(path);
        } catch (IOException e) {
            throw new KensaException(String.format("Unable to recreate directory [%s]", path), e);
        }
    }

    public static void copyResource(String url, Path destDir) {
        try {
            InputStream resourceStream = IoUtil.class.getResourceAsStream(url);
            if (resourceStream == null) {
                throw new KensaException(String.format("Did not find resource [%s]", url));
            }
            if (!Files.isDirectory(destDir)) {
                throw new KensaException(String.format("Destination [%s] is not a directory", destDir.toString()));
            }
            Files.copy(resourceStream, destDir.resolve(Paths.get(url).getFileName()));
        } catch (IOException e) {
            throw new KensaException(String.format("Unable to copy resource [%s] to [%s]", url, destDir), e);
        }
    }

    private static class DeleteAllVisitor extends SimpleFileVisitor<Path> {

        @Override
        public FileVisitResult visitFile(Path file, BasicFileAttributes attributes) throws IOException {
            Files.delete(file);
            return CONTINUE;
        }

        @Override
        public FileVisitResult postVisitDirectory(Path directory, IOException exception) throws IOException {
            Files.delete(directory);
            return CONTINUE;
        }
    }
}

File: src/main/java/dev/kensa/util/IoUtil.kt
package dev.kensa.util

import dev.kensa.KensaException
import java.io.IOException
import java.nio.file.*
import java.nio.file.FileVisitResult.CONTINUE
import java.nio.file.attribute.BasicFileAttributes

object IoUtil {
    private val DELETE_ALL_VISITOR = DeleteAllVisitor()

    fun recreate(path: Path) {
        try {
            if (Files.exists(path)) {
                Files.walkFileTree(path, DELETE_ALL_VISITOR)
            }
            Files.createDirectories(path)
        } catch (e: IOException) {
            throw KensaException("Unable to recreate directory [$path]", e)
        }
    }

    fun copyResource(url: String, destDir: Path) {
        try {
            val resourceStream = IoUtil::class.java.getResourceAsStream(url) ?: throw KensaException("Did not find resource [$url]")
            if (!Files.isDirectory(destDir)) {
                throw KensaException("Destination [$destDir] is not a directory")
            }
            Files.copy(resourceStream, destDir.resolve(Paths.get(url).fileName))
        } catch (e: IOException) {
            throw KensaException("Unable to copy resource [$url] to [$destDir]", e)
        }
    }

    private class DeleteAllVisitor : SimpleFileVisitor<Path>() {
        override fun visitFile(file: Path?, attrs: BasicFileAttributes?): FileVisitResult {
            file?.let {
                Files.delete(it)
            }
            return CONTINUE
        }

        override fun postVisitDirectory(directory: Path?, exc: IOException?): FileVisitResult {
            directory?.let {
                Files.delete(directory)
            }
            return CONTINUE
        }
    }
}